### ISSUE TITLE
Crash on start / Insufficient free space in range - Cannot relocate program.

### ISSUE THREAD

**Issue by danielstr007** [memtest86plus/memtest86plus#49]
![This is an image](https://user-images.githubusercontent.com/102787488/164422395-f32b34ea-78cd-4c21-b9cf-8ab80317b7b8.png)

Current release crash directly after start

With older release from 07.04. everything running fine:
![image2022-4-12_14-12-5](https://user-images.githubusercontent.com/102787488/164422890-7579a228-784e-4315-a487-c2809897d49f.png)

Any ideas?

**Comment by debrouxl** [comment]
Nice computer you have there :)

What's the size of the respective memtest86+ binaries from the previous version and this one ? I'd venture that the binary's size increased slightly, and .text + .data + .rodata + .bss do no longer fit in the 100-665c7 range.

**Comment by danielstr007** [comment]
Hey,
thanks yes, this is one of our larger systems 

I could only quick check the complied bin files, size increased - so this looks like the problem?

Current:
-rwxr-xr-x 1 145792 Apr 21 08:51 memtest-6beta/memtest86plus-main/build64/memtest.bin
-rw-r--r-- 1 144256 Apr 21 08:51 memtest-6beta/memtest86plus-main/build64/memtest_shared.bin
OLD:
-rwxr-xr-x 1 133216 Apr  7 10:54 memtest-6beta_old/memtest86plus-main/build64/memtest.bin
-rw-r--r-- 1 131680 Apr  7 10:54 memtest-6beta_old/memtest86plus-main/build64/memtest_shared.bin

**Comment by x86fr** [comment]
can you try to disable some CPU/Core? The root of the issue probably comes from the size of the AP structures and stack (394 kB total!), more than the size of the binary by itself.

**Comment by danielstr007** [comment]
> can you try to disable some CPU/Core? The root of the issue probably comes from the size of the AP structures and stack (394 kB total!), more than the size of the binary by itself.

Will try to disable all cores to 1 tomorrow, but in the older release it worked? 
So, it looks like there was a code change?

**Comment by x86fr** [comment]
Yes, but I think you were already extremely close to the limit. The tiny 12KB added code took it over the line.
Disabling only half of the cores (or even 1/4) should be enough.

Anyway, this is an issue we have to solve to claim support for 256 threads or even more.

**Comment by danielstr007** [comment]
![2022-04-21 13_24_25-Virtual Desktop - Desktop Viewer](https://user-images.githubusercontent.com/102787488/164448067-5f89b352-568d-4f56-afbd-3fb42b4752a1.png)
This is correct, booted with F2 Key -> SMP disabled
memtest running fine
Hope it can be fixed until the full release

**Comment by x86fr** [comment]
Sure! But the biggest issue is to test the fix because such awesome platform is not common. Your help will be appreciated :)

**Comment by danielstr007** [comment]
No problem, give me a bump - I will create a new build and test on such a box.

**Comment by martinwhitaker** [comment]
The problem with this system is that the BIOS has reserved a chunk of memory about 2/5 of the way through low memory, significantly reducing the amount of contiguous low memory available. I don't know how common that is - on my machines, pretty much all of low memory is free, leaving ample room to support 256 threads.

**Comment by debrouxl** [comment]
Merging something along the lines of #24 , which I've just rebased, may help slightly in the meantime, if the decrease allows going below an alignment threshold, otherwise it won't change anything.

**Comment by danielstr007** [comment]
> Merging something along the lines of #24 , which I've just rebased, may help slightly in the meantime, if the decrease allows going below an alignment threshold, otherwise it won't change anything.

Tried to build on this branch optimize_structs - build 55d9013ca4 
We saved a few kB 394->389
![2022-04-21 18_21_14-Virtual Desktop - Desktop Viewer](https://user-images.githubusercontent.com/102787488/164507438-16f79c9e-af46-4299-83ef-ce481920f73e.png)

But still crashed, if it want to start on all cores.

```
   text    data     bss     dec     hex filename
 132096       0       0  132096   20400 memtest-6beta_old/memtest86plus-main/build64/memtest.efi (old ~07.04. Build - OK)
 144384       0       0  144384   23400 memtest-6beta/memtest86plus-main/build64/memtest.efi (Current branch main 21.04.)
 139264       0       0  139264   22000 memtest-6beta_55d9013ca4/memtest86plus-55d9013ca4ae06a9e28fff1d793c3ccad655cd32/build64/memtest.efi (optimize_structs - build 55d9013ca4)
```
Booting this with manual F1 override setting 0-99 Cores, it starts working without crash:
![image](https://user-images.githubusercontent.com/102787488/164508409-15ebd8bd-7bfb-4383-a807-49474bde2856.png)



Found a small glitch, which might be also worth to fix:
![2022-04-21 18_30_44-Virtual Desktop - Desktop Viewer](https://user-images.githubusercontent.com/102787488/164508582-71f23b49-99ce-49e0-8dfb-db0a37ebb218.png)
You can "only" select up to 99 - in a decade of more than 100 cores, this might be worth to allow 3 digits as input.

**Comment by danielstr007** [comment]
![2022-04-25 08_35_44-Virtual Desktop - Desktop Viewer](https://user-images.githubusercontent.com/102787488/165033774-950dacd7-d75b-49cb-902b-262f236ac570.png)

Thanks, the memory size extension to 4MB works for me on this platform now very well. 
For me the issue can be closed as fixed.
